Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp

```diff
@@ -0,0 +1,309 @@
+ * along with this library; see the file COPYING.LIB.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "config.h"
+#include "RenderSVGResourceClipper.h"
+
+#include "ElementIterator.h"
+#include "Frame.h"
+#include "FrameView.h"
+#include "HitTestRequest.h"
+#include "HitTestResult.h"
+#include "IntRect.h"
+#include "Logging.h"
+#include "RenderElementInlines.h"
+#include "RenderLayer.h"
+#include "RenderLayerInlines.h"
+#include "RenderSVGResourceClipperInlines.h"
+#include "RenderSVGText.h"
+#include "RenderStyle.h"
+#include "RenderView.h"
+#include "SVGClipPathElement.h"
+#include "SVGContainerLayout.h"
+#include "SVGElementTypeHelpers.h"
+#include "SVGNames.h"
+#include "SVGPathData.h"
+#include "SVGRenderingContext.h"
+#include "SVGResources.h"
+#include "SVGResourcesCache.h"
+#include "SVGUseElement.h"
+#include <wtf/IsoMallocInlines.h>
+#include <wtf/SetForScope.h>
+#include <wtf/text/TextStream.h>
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(RenderSVGResourceContainer);
+
+RenderSVGResourceContainer::RenderSVGResourceContainer(SVGElement& element, RenderStyle&& style)
+    : RenderSVGHiddenContainer(Type::SVGResourceClipper, element, WTFMove(style))
+    , m_id(element.getIdAttribute())
+{
+}
+
+RenderSVGResourceContainer::~RenderSVGResourceContainer() = default;
+
+void RenderSVGResourceContainer::willBeDestroyed()
+{
+    if (m_registered) {
+        treeScopeForSVGReferences().removeSVGResource(m_id);
+        m_registered = false;
+    }
+
+    RenderSVGHiddenContainer::willBeDestroyed();
+}
+
+void RenderSVGResourceContainer::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle)
+{
+    RenderSVGHiddenContainer::styleDidChange(diff, oldStyle);
+
+    if (!m_registered) {
+        m_registered = true;
+        registerResource();
+    }
+}
+
+void RenderSVGResourceContainer::idChanged()
+{
+    // Invalidate all our current clients.
+    removeAllClientsFromCache();
+
+    // Remove old id, that is guaranteed to be present in cache.
+    treeScopeForSVGReferences().removeSVGResource(m_id);
+    m_id = element().getIdAttribute();
+
+    registerResource();
+}
+
+void RenderSVGResourceContainer::registerResource()
+{
+    auto& treeScope = this->treeScopeForSVGReferences();
+    if (!treeScope.isIdOfPendingSVGResource(m_id)) {
+        treeScope.addSVGResource(m_id, *this);
+        return;
+    }
+
+    auto elements = copyToVectorOf<Ref<SVGElement>>(treeScope.removePendingSVGResource(m_id));
+
+    treeScope.addSVGResource(m_id, *this);
+
+}
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(RenderSVGResourceClipper);
+
+RenderSVGResourceClipper::RenderSVGResourceClipper(SVGClipPathElement& element, RenderStyle&& style)
+    : RenderSVGResourceContainer(element, WTFMove(style))
+{
+}
+
+RenderSVGResourceClipper::~RenderSVGResourceClipper() = default;
+
+SVGGraphicsElement* RenderSVGResourceClipper::shouldApplyPathClipping() const
```

Comment:
This could live in the DOM, aka. SVGClipPathElement?

Reply:
I moved it there, but maybe not so nice that SVGClipPathElement::shouldApplyPathClipping checks style as the first thing...